snowflake/datadog_checks/snowflake/check.py

```diff
@@ -0,0 +1,119 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from contextlib import closing
+
+import snowflake.connector as sf
+
+from datadog_checks.base import AgentCheck, ConfigurationError
+from datadog_checks.base.utils.db import QueryManager
+
+from . import queries
+from .config import Config
+
+METRIC_GROUPS = {
+    'snowflake.query': [queries.WarehouseLoad, queries.QueryHistory],
+    'snowflake.billing': [queries.CreditUsage, queries.WarehouseCreditUsage],
+    'snowflake.storage': [queries.StorageUsageMetrics],
+    'snowflake.storage.database': [queries.DatabaseStorageMetrics],
+    'snowflake.storage.table': [queries.TableStorage],
+    'snowflake.logins': [queries.LoginMetrics],
+    'snowflake.data_transfer': [queries.DataTransferHistory],
+    'snowflake.auto_recluster': [queries.AutoReclusterHistory],
+    'snowflake.pipe': [queries.PipeHistory],
+    'snowflake.replication': [queries.ReplicationUsage],
+}
+
+
+class SnowflakeCheck(AgentCheck):
+    """
+    Collect Snowflake account usage metrics
+    """
+
+    __NAMESPACE__ = 'snowflake'
+
+    SERVICE_CHECK_CONNECT = 'snowflake.can_connect'
+
+    def __init__(self, *args, **kwargs):
+        super(SnowflakeCheck, self).__init__(*args, **kwargs)
+        self.config = Config(self.instance)
+        self._conn = None
+
+        # Add default tags like account to all metrics
+        self._tags = self.config.tags + ['account:{}'.format(self.config.account)]
+
+        if self.config.password:
+            self.register_secret(self.config.password)
+
+        self.metric_queries = []
+        self.errors = []
+        for mgroup in self.config.metric_groups:
+            try:
+                self.metric_queries.extend(METRIC_GROUPS[mgroup])
+            except KeyError:
+                self.errors.append(mgroup)
+
+        if self.errors:
+            self.log.warning('Invalid metric_groups found in snowflake conf.yaml: %s', (', '.join(self.errors)))
+        if not self.metric_queries:
+            raise ConfigurationError('No valid metric_groups configured, please list at least one.')
+
+        self._query_manager = QueryManager(self, self.execute_query_raw, queries=self.metric_queries, tags=self._tags)
+        self.check_initializations.append(self._query_manager.compile_queries)
+
+    def check(self, _):
+        self.connect()
+
+        # Execute queries
+        self._query_manager.execute()
+
+        self._collect_version()
+
+    def execute_query_raw(self, query):
+        """
+        Executes query with timestamp from parts if comparing start_time field.
+        """
+        with closing(self._conn.cursor()) as cursor:
+            cursor.execute(query)
+
+            if cursor.rowcount is None or cursor.rowcount < 1:
+                self.log.debug("Failed to fetch records from query: `%s`", query)
+                return []
+            return cursor.fetchall()
+
+    def connect(self):
```

Comment:
We don't handle refreshing the connection in case something goes wrong. This is an important thing to add.

Reply:
addressed in connect()